libsolidity/analysis/TypeChecker.cpp

```diff
@@ -1740,10 +1769,12 @@ bool TypeChecker::visit(UnaryOperation const& _operation)
 			m_errorReporter.typeError(4907_error, _operation.location(), description);
 		_operation.annotation().type = subExprType;
 	}
-	else
-		_operation.annotation().type = result.get();
+
 	_operation.annotation().isConstant = false;
-	_operation.annotation().isPure = !modifying && *_operation.subExpression().annotation().isPure;
+	_operation.annotation().isPure =
+		!modifying &&
+		*_operation.subExpression().annotation().isPure &&
+		(!userDefinedFunctionType || userDefinedFunctionType->isPure());
```

Comment:
ok, I think I'm starting to better understand how you see it. So your approach is basically "if it's broken and `userDefinedOperator()` does not find it, ignore it, it must be a built-in operator"?

> Especially, when the opator is not allowed to be customized we can cleary tell that built-in operator is in charge, can't we?

I guess that's where we look at it differently. My view was that we're still inside analysis and the code is potentially very broken. The fact that we did not manage to match the definition does not mean it wasn't meant to be a custom operator and assuming that prematurely may lead to issuing confusing errors later in the analysis.

> does not perform any checks on type but does the filtering

The thing is - some of the things you're filtering out can only happen in broken code. I think the place of these filters is as checks here and the function should be able to assume that at the point where it's called, they cannot happen.

You have a point about unary vs binary. This is one filter I'd leave inside the function.

Reply:
Sure:
1. Custom operator used with a type that has no definition.
2. Expression with a custom operator that is not one of the customizable operators.
3. Operator having no definition or more than one definition.

Once you get past the type checker (so that all `using for`s have been visited and unary/binary operator use type-checked), none of these can happen. You could artificially trigger (1) or (2) by calling the function with contrived arguments but that's not useful and the function can block it with an assert without preventing us from doing anything useful.